iothub_client/tests/iothubtransport_mqtt_common_int_ut/iothubtransport_mqtt_common_int_ut.c

```diff
@@ -0,0 +1,1172 @@
+TEST_EXPECTED_MESSAGE_PROPERTIES creationTimeExpected = { NULL, NULL, NULL, NULL, NULL, NULL, NULL, TEST_CREATION_TIME_VALUE, NULL, NULL };
+
+#define TEST_ALL_SYSTEM_PROPERTIES "%24.cid=" TEST_CORRELATION_PROPERTY "&%24.uid=" TEST_MSG_USER_ID_VALUE \
+                                   "&%24.mid=" TEST_MSG_ID_VALUE "&%24.ct=" TEST_CONTENT_TYPE_VALUE \
+                                   "&%24.ce=" TEST_CONTENT_ENCODING_VALUE   "&%24.cdid="  TEST_CONNECTION_DEVICE_VALUE  \
+                                   "&%24.cmid=" TEST_CONNECTION_MODULE_VALUE "&%24.ctime=" TEST_CREATION_TIME_VALUE
+
+static const char* TEST_MQTT_MSG_ALL_SYSTEM_TOPIC = "devices/myDeviceId/messages/devicebound/" TEST_ALL_SYSTEM_PROPERTIES;
+TEST_EXPECTED_MESSAGE_PROPERTIES allSystemPropertiesSet1Expected = { TEST_CONTENT_TYPE_VALUE, TEST_CONTENT_ENCODING_VALUE, TEST_MSG_ID_VALUE, TEST_CORRELATION_PROPERTY, NULL, TEST_CONNECTION_MODULE_VALUE, TEST_CONNECTION_DEVICE_VALUE, TEST_CREATION_TIME_VALUE, TEST_MSG_USER_ID_VALUE, NULL};
+
+#define TEST_TOPICS_TO_IGNORE "iothub-operation=valueToIgnore&iothub-ack=valueToIgnore&%24.to=valueToIgnore&%24.on=valueToIgnore&%24.exp=valueToIgnore&devices/=valueToIgnore"
+#define TEST_TOPICS_TO_NOT_IGNORE "&devices=valueToApp1&to=valueToApp2&exp=valueToApp3&on=valueToApp4"
+static const char* TEST_MQTT_IGNORED_TOPICS= "devices/myDeviceId/messages/devicebound/" TEST_TOPICS_TO_IGNORE TEST_TOPICS_TO_NOT_IGNORE;
+
+char* expectedNotIgnoredKeys[] = {"devices", "to", "exp", "on"};
+char* expectedNotIgnoredValues[] = {"valueToApp1", "valueToApp2", "valueToApp3", "valueToApp4" };
+TEST_EXPECTED_APPLICATION_PROPERTIES expectedNotIgnored = { expectedNotIgnoredKeys, expectedNotIgnoredValues, 4};
+TEST_EXPECTED_MESSAGE_PROPERTIES mostlyIgnoredPropertiesExpected = { NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, &expectedNotIgnored};
+
+static const char* mqttNoMatchTopic[] = {
+    "",
+    "ThisIsNotCloseToBeingALegalTopic",
+    "/device/",
+    "devices/",
+    "devices/myDeviceId/messages",
+    "devices/myDeviceId/messages/deviceboun",
+    "/devices/myDeviceId/messages/devicebound",
+    // These are legal topics but as we're not subscribed to them they should be ignored.
+    "$iothub/twin/twinData",
+    "iothub/methods/methodData",
+    "devices/myDeviceId/modules/thisIsModuleID/inputs/" TEST_INPUT_QUEUE_1 "/"
+};
+static const size_t mqttNoMatchTopicLength = sizeof(mqttNoMatchTopic) / sizeof(mqttNoMatchTopic[0]);
+
+static const char* emptyPropertyMQTTTopics[] = {
+    "devices/myDeviceId/messages/devicebound/",
+    "devices/myDeviceId/messages/devicebound/&",
+    "devices/myDeviceId/messages/devicebound/&&",
+    "devices/myDeviceId/messages/devicebound/&&&",
+    "devices/myDeviceId/messages/devicebound/=",
+    "devices/myDeviceId/messages/devicebound/fooBar",
+    //"devices/myDeviceId/messages/devicebound/==",
```

Comment:
needed?

Reply:
It didn't belong there since we end up with a name="" and value="=" on this case, which is really weird.  Mostly test should be there to make sure SDK doesn't crash with all these what turn into 0 len strings thrown at us.  Adding a separate test case for this.